src/pytorch_lightning/plugins/environments/azure_openmpi_environment.py

```diff
@@ -0,0 +1,77 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.plugins.environments.lightning_environment import find_free_network_port
+
+
+class AzureOpenMPIEnvironment(ClusterEnvironment):
+    """Environment for an OpenMPI environment on Azure.
+
+    See Azure documentation:
+        - https://docs.microsoft.com/en-us/azure/machine-learning/how-to-train-distributed-gpu#mpi
+        - https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/batch/batch-compute-node-environment-variables.md
+    """
+
+    def __init__(self, devices: int = 1) -> None:
+        """devices : devices per node (same as trainer parameter)"""
+        super().__init__()
+        self.devices = devices
+
+    @property
+    def creates_processes_externally(self) -> bool:
+        """Return True if the cluster is managed (you don't launch processes yourself)"""
+        return True
+
+    @property
+    def main_address(self) -> str:
+        # AZ_BATCH_MASTER_NODE should be defined when num_nodes > 1
+        if "AZ_BATCH_MASTER_NODE" in os.environ:
+            return os.environ.get("AZ_BATCH_MASTER_NODE").split(":")[0]
+        elif "AZ_BATCHAI_MPI_MASTER_NODE" in os.environ:
+            return os.environ.get("AZ_BATCHAI_MPI_MASTER_NODE")
+        else:
+            return os.environ.get("MASTER_ADDR", "127.0.0.1")
+
+    @property
+    def main_port(self) -> int:
+        # AZ_BATCH_MASTER_NODE should be defined when num_nodes > 1
+        if "AZ_BATCH_MASTER_NODE" in os.environ:
+            return int(os.environ.get("AZ_BATCH_MASTER_NODE").split(":")[1])
+        else:
+            return int(os.environ.get("MASTER_PORT", find_free_network_port()))
```

Comment:
Need to be careful here: If each process calls `find_free_network_port()`, all processes will end up with a different "generated" port. Maybe you took this code from our LightningEnvironment, but it is important to note that we do it there only because we know we call it on the main process first, and then assign 

```py
os.environ["MASTER_PORT"] = cluster_env.main_port
```

Here, this won't be the case because the assumption is that processes get created externally.

Reply:
This could explain why you got
`RuntimeError: The server socket has failed to listen on any local network address. The server socket has failed to bind to [::]:6105 (errno: 98 - Address already in use). The server socket has failed to bind to ?UNKNOWN? (errno: 98 - Address already in use).`

Reply:
@jessecambon What if we hardcode a port that we know is available. Could also be made configurable as a parameter input to the `__init__`

Reply:
@awaelchli thanks, that makes sense. I removed `find_free_network_port()` and changed the code back to first look in the `MASTER_PORT` environmental variable and then default to a hardcoded port value. For some reason I still get these errors whether I set `MASTER_PORT` or let it use the hardcoded port value:

```sh
RuntimeError: The server socket has failed to listen on any local network address. The server socket has failed to bind to [::]:57345 (errno: 98 - Address already in use). The server socket has failed to bind to ?UNKNOWN? (errno: 98 - Address already in use).
```